File: services/frontend/src/lib/notificationLinks.ts
import type { UiMode } from '@/hooks/useResolvedUiMode'

/**
 * Deep links for the "new grading on your submission" notifications.
 *
 * Mirrors the backend helper that builds the link in the notification email
 * (services/shared/mailer/notification_links.py), so the bell, the
 * notifications page and the email all open the same place.
 */
export const EVALUATION_RECEIVED_TYPES = [
  'evaluation_received_human',
  'evaluation_received_immediate',
  'evaluation_received_batch',
] as const

export type EvaluationReceivedType = (typeof EVALUATION_RECEIVED_TYPES)[number]

export function isEvaluationReceivedType(
  type: string | null | undefined,
): type is EvaluationReceivedType {
  return (EVALUATION_RECEIVED_TYPES as readonly string[]).includes(type ?? '')
}

/**
 * A student opens an exam in the student shell. Everyone else lands on the
 * project's own task list, where the grading opens per task. Returns null
 * when the notification carries no project.
 */
export function getEvaluationReceivedHref(
  data: Record<string, any> | null | undefined,
  uiMode: UiMode,
): string | null {
  const projectId = data?.project_id
  if (!projectId) return null
  if (uiMode === 'student' && data?.project_kind === 'exam') {
    return `/student/exams/${projectId}`
  }
  return `/projects/${projectId}/my-tasks`
}

File: services/shared/mailer/notification_links.py
"""Where a notification points in the web app.

The email path uses this to build an absolute link for the recipient's brand.
The frontend mirrors it in src/lib/notificationLinks.ts to route the in-app
notification by the viewer's UI mode. Keep the two in step.
"""

from collections.abc import Mapping
from typing import Any

from models import NotificationType

EVALUATION_RECEIVED_TYPES = frozenset(
    {
        NotificationType.EVALUATION_RECEIVED_HUMAN.value,
        NotificationType.EVALUATION_RECEIVED_IMMEDIATE.value,
        NotificationType.EVALUATION_RECEIVED_BATCH.value,
    }
)


def is_evaluation_received_type(notification_type: Any) -> bool:
    """True for the three "new grading on your submission" types (enum or value)."""
    value = getattr(notification_type, "value", notification_type)
    return isinstance(value, str) and value in EVALUATION_RECEIVED_TYPES


def evaluation_received_path(
    data: Mapping[str, Any] | None, student_surface: bool
) -> str | None:
    """Path of the page that shows a new grading to its annotator.

    A student reads an exam's gradings on the exam page of the student
    surface. Everyone else, and every non-exam project, uses the project's
    My Tasks list, which opens the submission with its scores. None when the
    notification names no project.
    """
    data = data or {}
    project_id = data.get("project_id")
    if not project_id:
        return None
    if student_surface and data.get("project_kind") == "exam":
        return f"/student/exams/{project_id}"
    return f"/projects/{project_id}/my-tasks"
